### PR TITLE
[FLINK-26738]Mark StateDescriptor#defaultValue as deprecated and add doc

### PR DESCRIPTION
**Pull request by Mulavar**
## What is the purpose of the change

Mark StateDescriptor#defaultValue as deprecated and add doc/javadoc.

## Verifying this change

This change is a trivial rework / code cleanup without any test coverage.

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): (no)
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: (no)
  - The serializers: (no)
  - The runtime per-record code paths (performance sensitive): (no)
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Kubernetes/Yarn, ZooKeeper: (no)
  - The S3 file system connector: (no)

## Documentation

  - Does this pull request introduce a new feature? (no)
  - If yes, how is the feature documented? (docs / JavaDocs )

### CODE REVIEW

File: docs/content/docs/dev/datastream/fault-tolerance/state.md [hunk 1]
```diff
@@ -432,6 +432,8 @@ will lead to compatibility failure and `StateMigrationException`.
 - The map state with TTL currently supports null user values only if the user value serializer can handle null values. 
 If the serializer does not support null values, it can be wrapped with `NullableSerializer` at the cost of an extra byte in the serialized form.
 
+- If enable state TTL configuration, the default value configuration in `StateDescriptor` will no longer take an effect.
```
Comment (Myasuka) [inline docs/content/docs/dev/datastream/fault-tolerance/state.md:435]: I think it's better to explain why this operation is reasonable.

File: flink-core/src/main/java/org/apache/flink/api/common/state/StateDescriptor.java [hunk 2]
```diff
@@ -258,7 +258,7 @@ public boolean isQueryable() {
      * Configures optional activation of state time-to-live (TTL).
      *
      * <p>State user value will expire, become unavailable and be cleaned up in storage depending on
-     * configured {@link StateTtlConfig}.
+     * configured {@link StateTtlConfig}, and the field defaultValue will be invalid.
```
Comment (Myasuka) [inline flink-core/src/main/java/org/apache/flink/api/common/state/StateDescriptor.java:261]: I think the explanation of the change of `defaultValue` could be placed in another statement starts with `<p>`.

File: flink-core/src/main/java/org/apache/flink/api/common/state/StateDescriptor.java [hunk 3]
```diff
@@ -104,7 +104,7 @@
     @Nonnull private StateTtlConfig ttlConfig = StateTtlConfig.DISABLED;
 
     /** The default value returned by the state when no other value is bound to a key. */
```
Comment (Myasuka) [inline flink-core/src/main/java/org/apache/flink/api/common/state/StateDescriptor.java:106]: I think we'd better to describe why this value is deprecated.

**Review by Myasuka — CHANGES_REQUESTED** [review]
Thanks for creating this PR.
I noticed that the CI is also broken, maybe we need to rebase with latest flink master branch and force push update this PR.

**Comment by Myasuka** [comment]
Before another turn of review. Please:

1. Fill up the description of this PR.

**Comment by Mulavar** [comment]
> Before another round of review. Please:
> 
> 1. Fill up the description of this PR.
> 2. Resolve the broken CI due to checkstyle.

Thanks, done.

**Comment by Mulavar** [comment]
This is the first time I push a pr to the Flink and I do not know how to re-run the CI, could you give me some help, thanks.

**Comment by Myasuka** [comment]
> This is the first time I push a pr to the Flink and I do not know how to re-run the CI, could you give me some help, thanks.

In general, once you udapted the branch, the CI would pend to trigger. And you can also manually trigger it via "Bot commands".

File: flink-core/src/main/java/org/apache/flink/api/common/state/StateDescriptor.java [hunk 4]
```diff
@@ -103,8 +103,13 @@
     /** The configuration of state time-to-live(TTL), it is disabled by default. */
     @Nonnull private StateTtlConfig ttlConfig = StateTtlConfig.DISABLED;
 
-    /** The default value returned by the state when no other value is bound to a key. */
-    @Nullable protected transient T defaultValue;
+    /**
+     * The default value returned by the state when no other value is bound to a key.
+     *
+     * @deprecated since user cannot judge whether the returned default value is for the real
+     *     current state value or just itself, which may make the semantics unclear.
```
Comment (Myasuka) [inline flink-core/src/main/java/org/apache/flink/api/common/state/StateDescriptor.java:110]: I think the description might be better:
`To make the semantics more clear, user should manually manage the default value if the contents of the state is {@code null}.`

File: docs/content/docs/dev/datastream/fault-tolerance/state.md [hunk 5]
```diff
@@ -432,6 +432,8 @@ will lead to compatibility failure and `StateMigrationException`.
 - The map state with TTL currently supports null user values only if the user value serializer can handle null values. 
 If the serializer does not support null values, it can be wrapped with `NullableSerializer` at the cost of an extra byte in the serialized form.
 
+- If enable state TTL configuration, the default value configuration in `StateDescriptor` will no longer take an effect, since if we get the value that equals with the default value from TTL state, we can not judge whether the key is expired, which make the semantics unclear.
```
Comment (Myasuka) [inline docs/content/docs/dev/datastream/fault-tolerance/state.md:435]: How about change like this?
```suggestion
- With TTL enabled configuration, the `defaultValue` in `StateDescriptor`, which is atucally already deprecated, will no longer take an effect. This aims to make the semantics more clear and let user manually manage the default value if the contents of the state is null or expired.
```

File: docs/content.zh/docs/dev/datastream/fault-tolerance/state.md [hunk 6]
```diff
@@ -383,6 +383,8 @@ Heap state backend 会额外存储一个包括用户状态以及时间戳的 Jav
 
 - 当前开启 TTL 的 map state 仅在用户值序列化器支持 null 的情况下，才支持用户值为 null。如果用户值序列化器不支持 null，
 可以用 `NullableSerializer` 包装一层。
+  
+- 如果启用 TTL 配置，则之前创建 `StateDescriptor` 设置的 `defaultValue` 将会失效，因为如果从 TTL 的状态中拿到了默认值，用户无法判断这个值是尚未过期的状态值还是状态过期后拿到的默认值，从而也无法判断状态是否过期，这将导致语义混乱。
```
Comment (Myasuka) [inline docs/content.zh/docs/dev/datastream/fault-tolerance/state.md:387]: I think we should update the chinese transformation with English part updated.

**Review by Myasuka — CHANGES_REQUESTED** [review]
Thanks for the update!
Apart from taking a look at my comments, please also change your commit message to 
`[FLINK-26738][state] Mark StateDescriptor#defaultValue as deprecated with docs`.

Moreover, I noticed that the original commit author is from `dongjianhui03@meituan.com` which is not binding with your github account `Mulavar`.

File: docs/content.zh/docs/dev/datastream/fault-tolerance/state.md (hunk 6 above)
Comment (Mulavar) [reply docs/content.zh/docs/dev/datastream/fault-tolerance/state.md]: thanks! done!

**Comment by Mulavar** [comment]
@flinkbot run azure

**Review by Myasuka — COMMENTED** [review]
The CI is still broken, maybe you can rebase with latest master branch again.

**Review by Myasuka — APPROVED** [review]
LGTM.
I'll manually merge your commit later with new message:
`[FLINK-26738][state] Mark StateDescriptor#defaultValue as deprecated with clear docs`.

**Pull request closed without merging**